File: java/cours/src/session2/td/agregation/NatDeleguantEtat.java
package session2.td.agregation;

public abstract class NatDeleguantEtat implements Nat{
	
	private EtatNaturelPur etat;
	
	public NatDeleguantEtat(EtatNaturelPur etat) {
		this.etat=etat;
	}

	public EtatNaturelPur etat() {
		return this.etat;
	}
	
	public int val() {
		return this.etat.val();
	}
	
	public Nat predecesseur() {
		return creerNatAvecEtat(etat.predecesseur());
	}
	
	public int chiffre(int val) {
		return this.etat.chiffre(val);
	}
	
	public int taille() {
		return this.etat.taille();
	}
	
	public Nat creerNatAvecValeur(int val) {
		return creerNatAvecEtat(this.etat.creerNatAvecValeur(val));
	}
	
	public Nat creerZero() {
		return creerNatAvecEtat(this.etat.creerZero());
	}
	
	public Nat creerSuccesseur(Nat pred) {
		return creerNatAvecEtat(this.etat.creerSuccesseur(pred.etat()));
	}
	
	public Nat creerNatAvecRepresentation(String repDecimale) {
		return creerNatAvecEtat(this.etat.creerNatAvecRepresentation(repDecimale));
	}
	
	public boolean estNul() {
		if(this.equals(this.zero())) return true;
		return false;
	}
	
}

File: typescript/session2/agregation/NatDeleguantEtat.ts
import {EtatNaturelPur} from "./EtatNaturelIPur";
import {Nat} from "./Nat";

export abstract class NatDeleguantEtat implements Nat{
    abstract creerNatAvecEtat(etat: EtatNaturelPur): Nat;
    abstract somme(x: Nat): Nat;
    abstract zero(): Nat;
    abstract produit(x: Nat): Nat;
    abstract un(): Nat;
    abstract modulo(x: Nat): Nat;
    abstract div(x: Nat): Nat;
    abstract creerNatValue(value: number): Nat;
    abstract creerNatString(value: string): Nat;
    abstract creerNatSucc(pred: Nat): Nat;
    abstract creerNatZero(): Nat;

    private etat : EtatNaturelPur;

    public NatDeleguantEtat(etat : EtatNaturelPur) {
        this.etat=etat;
    }

    getEtat() : EtatNaturelPur {
        return this.etat;
    }

    val() : number {
        return this.etat.val();
    }

    predecesseur() : Nat {
        return this.creerNatAvecEtat(this.etat.predecesseur());
    }

    chiffre(val : number) : number {
        return this.etat.chiffre(val);
    }

    taille() : number {
        return this.etat.taille();
    }

    creerNatAvecValeur(val : number) : Nat{
        return this.creerNatAvecEtat(this.etat.creerNatValue(val));
    }

    creerZero() : Nat {
        return this.creerNatAvecEtat(this.etat.creerNatZero());
    }

    creerSuccesseur(pred : Nat) : Nat {
        return this.creerNatAvecEtat(this.etat.creerNatSucc(pred.getEtat()));
    }

    creerNatAvecRepresentation(repDecimale : string) : Nat {
        return this.creerNatAvecEtat(this.etat.creerNatString(repDecimale));
    }

    estNul() : boolean {
        return this.val() == 0;
    }
}
